### PR TITLE
Enable training with pre-training mode

### PR DESCRIPTION
**Pull request by RobotSail**
This PR enables training in pre-training mode from a fixed dataset (non-streamed). When a user enables pretraining mode by specifying the `pretraining_config` argument, the training code handles the dataset by packing all input IDs into a dense contiguous list of samples and chunking them up into segments of length  `pretraining_config.block_size` . 

Internally, a new `PretrainingBlockDataset` is added which allows us to handle the samples as dense chunks. 

When using pretraining mode, we only expect a dataset which contains an `input_ids` column in each listing, where each record is NOT expected to have its lengths match `block_size`. Instead, this chunking happens in the actual training code inside of the `PretrainingBlockDataset` logic.

## Summary by CodeRabbit

* **New Features**
  * Pretraining Mode: --block-size CLI/API, PretrainingConfig, block-based dataset and padded minibatch collation, pretraining-aware data loaders.

* **Improvements**
  * Deterministic startup seeding, tokenizer↔model pad-token alignment, preserved RNG for checkpoints, AdamW optimizer parameters exposed, refined loss scaling and logging.

* **Documentation**
  * Added Pretraining Mode docs and examples in README.

* **Tests**
  * New comprehensive tests for pretraining dataset and data-loader; updated related tests.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by RobotSail** [comment]
@NikhilNayak-debug Yes, in continual pre-training we essentially just continue to pretrain on top of an already trained/fine-tuned model in order to inject more data. It turns out that this doesn't actually break existing instruction capability, making it a great method of separating the knowledge injection and instruct-tuning phases.
